bindinfo/bind_cache.go

```diff
@@ -0,0 +1,213 @@
+func (c *bindCache) get(key bindCacheKey) []*BindRecord {
+	value, hit := c.cache.Get(key)
+	if !hit {
+		return nil
+	}
+	typedValue := value.([]*BindRecord)
+	return typedValue
+}
+
+// set inserts an item to the cache. It's not thread-safe.
+// Only other functions of the bindCache can use this function.
+func (c *bindCache) set(key bindCacheKey, value []*BindRecord) bool {
+	mem := calcBindCacheKVMem(key, value)
+	if mem > c.memCapacity { // ignore this kv pair if its size is too large
+		return false
+	}
+	bindRecords := c.get(key)
+	if bindRecords != nil {
+		// Remove the origin key-value pair.
+		mem -= calcBindCacheKVMem(key, bindRecords)
+	}
+	for mem+c.memTracker.BytesConsumed() > c.memCapacity {
+		evictedKey, evictedValue, evicted := c.cache.RemoveOldest()
+		if !evicted {
+			return false
+		}
+		c.memTracker.Consume(-calcBindCacheKVMem(evictedKey.(bindCacheKey), evictedValue.([]*BindRecord)))
+	}
+	c.memTracker.Consume(mem)
+	c.cache.Put(key, value)
+	return true
+}
+
+// delete remove an item from the cache. It's not thread-safe.
+// Only other functions of the bindCache can use this function.
+func (c *bindCache) delete(key bindCacheKey) bool {
+	bindRecords := c.get(key)
+	if bindRecords != nil {
+		mem := calcBindCacheKVMem(key, bindRecords)
+		c.cache.Delete(key)
+		c.memTracker.Consume(-mem)
+	}
+	return true
+}
+
+// GetBindRecord gets the BindRecord from the cache.
+// The return value is not read-only, but it shouldn't be changed in the caller functions.
+// The function is thread-safe.
+func (c *bindCache) GetBindRecord(hash, normdOrigSQL, db string) *BindRecord {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	bindRecords := c.get(bindCacheKey(hash))
+	for _, bindRecord := range bindRecords {
+		if bindRecord.OriginalSQL == normdOrigSQL {
+			return bindRecord
+		}
+	}
+	return nil
+}
+
+// GetAllBindRecords return all the bindRecords from the bindCache.
+// The return value is not read-only, but it shouldn't be changed in the caller functions.
+// The function is thread-safe.
+func (c *bindCache) GetAllBindRecords() []*BindRecord {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	values := c.cache.Values()
+	var bindRecords []*BindRecord
+	for _, vals := range values {
+		bindRecords = append(bindRecords, vals.([]*BindRecord)...)
+	}
+	return bindRecords
+}
+
+// SetBindRecord sets the BindRecord to the cache.
+// The function is thread-safe.
+func (c *bindCache) SetBindRecord(hash string, meta *BindRecord) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	cacheKey := bindCacheKey(hash)
+	metas := c.get(cacheKey)
+	for i := range metas {
+		if metas[i].OriginalSQL == meta.OriginalSQL {
+			metas[i] = meta
+		}
+	}
+	c.set(cacheKey, []*BindRecord{meta})
+}
+
+// RemoveDeletedBindRecord removes the BindRecord which has same originSQL with specified BindRecord.
+// The function is thread-safe.
+func (c *bindCache) RemoveDeletedBindRecord(hash string, meta *BindRecord) {
```

Comment:
How about using `RemoveBindRecord`? It's just a cache, shouldn't be aware of binding status.